File: round_robin_lb.go
package main

import (
	"sync"
	"testing"
)

type LoadBalancer struct {
	servers      []string
	currentIndex int
	mu           sync.Mutex
}

// NewLoadBalancer creates a new LoadBalancer instance from the given list of servers.
//
// The input slice is copied, so it can be modified independently of the LoadBalancer.
func NewLoadBalancer(servers []string) *LoadBalancer {
	copiedServers := make([]string, len(servers))
	copy(copiedServers, servers)
	return &LoadBalancer{
		servers:      copiedServers,
		currentIndex: 0,
	}
}

// GetNextServer returns the next server in the round-robin sequence.
// It locks the LoadBalancer to ensure thread safety and updates the current index
// to maintain the order of server selection.

func (lb *LoadBalancer) GetNextServer() string {
	lb.mu.Lock()
	defer lb.mu.Unlock()

	server := lb.servers[lb.currentIndex]
	lb.currentIndex = (lb.currentIndex + 1) % len(lb.servers)
	return server
}

// TestRoundRobin verifies that the LoadBalancer correctly rotates through the servers
// in a round-robin sequence.
func TestRoundRobin(t *testing.T) {
	serverList := []string{"Server1", "Server2", "Server3"}
	lb := NewLoadBalancer(serverList)

	expected := []string{"Server1", "Server2", "Server3", "Server1", "Server2"}
	for i := 0; i < 5; i++ {
		server := lb.GetNextServer()
		if server != expected[i] {
			t.Errorf("Expected %s, got %s at iteration %d", expected[i], server, i)
		}
	}
}

File: round_robin_lb.py
class LoadBalancer:
    def __init__(self, servers):
        """
        Create a new LoadBalancer from the given list of servers.

        The input list is copied, so it can be modified independently of the
        LoadBalancer.
        """
        self.servers = list(servers)
        self.current_index = 0

    def get_next_server(self):
        """
        Returns the next server in the round-robin sequence.

        This method updates the current index to maintain the order of 
        server selection in a thread-safe manner.
        """

        next_server = self.servers[self.current_index]
        self.current_index = (self.current_index + 1) % len(self.servers)
        return next_server

if __name__ == "__main__":
    server_list = ["Server1", "Server2", "Server3"]
    load_balancer = LoadBalancer(server_list)

    for i in range(10):
        next_server = load_balancer.get_next_server()
        print(f"Request {i+1}: Routed to {next_server}")